source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,134 @@ namespace aspect
+    const unsigned int n_q_points_21 = quadrature_formula_21.size();
+    FEValues<dim> fe_values_12 (mapping,
+                                finite_element,
+                                quadrature_formula_12,
+                                update_values   |
+                                update_quadrature_points |
+                                update_JxW_values);
+    std::vector<double> compositional_values_12(n_q_points_12);
+    FEValues<dim> fe_values_21 (mapping,
+                                finite_element,
+                                quadrature_formula_21,
+                                update_values   |
+                                update_quadrature_points |
+                                update_JxW_values);
+    std::vector<double> compositional_values_21(n_q_points_21);
+
+    std::vector<double> local_compositional_average_integrals (parameters.n_compositional_fields);
+
+    const double composition_exact_max = 1.0;//TODO: need to add into the parameters list, let the user to define.
+    const double composition_exact_min = 0;//TODO: same as comment above
+
+    LinearAlgebra::BlockVector distributed_vector (introspection.index_sets.system_partitioning,
+                                                   mpi_communicator);
+    distributed_vector = vector;
+
+    std::vector<types::global_dof_index> local_dof_indices (finite_element.dofs_per_cell);
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell != endc; ++cell)
+      {
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (local_dof_indices);
+            fe_values_12.reinit (cell);
+            fe_values_21.reinit (cell);
+
+            double local_area_integral = 0;
+
+            for (unsigned int q = 0; q < n_q_points_12; ++q)
+              {
+                local_area_integral += fe_values_12.JxW(q);
+              }
+            for (unsigned int c = 0; c < parameters.n_compositional_fields; ++c)
+              {
+                fe_values_12[introspection.extractors.compositional_fields[c]].get_function_values(vector,
+                    compositional_values_12);
+                fe_values_21[introspection.extractors.compositional_fields[c]].get_function_values(vector,
+                    compositional_values_21);
+
+                local_compositional_average_integrals[c] = 0;
+
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_compositional_average_integrals[c] += compositional_values_12[q]*fe_values_12.JxW(q);
+                  }
+                local_compositional_average_integrals[c] /= local_area_integral;
+
+                double min_composition=compositional_values_12[0];
+                double max_composition=compositional_values_12[0];
+                for (unsigned int q = 1; q < n_q_points_12; ++q)
```

Comment:
Ah, I understand now, you initialize the variables with the first value so you do not need to check it again. Still, I think it would be easier to understand if you could initialize the variables with `std::numeric_limits<double>::max()` and  `-std::numeric_limits<double>::max()` and just check every value. The slowdown would be negligible (one comparison is nothing compared to the total cost) and it would make the code easier to understand.